README.md
# Blinds

## About Blinds

Blinds is a Javascript script used to easily add dark mode styles to a website. We believe giving users the option for dark mode is no longer negotiable, and this applies to every type of website, in every niche. Blinds takes the pain out of development by making the process as easy and intuitive as possile. Some of the features include:

- Simple plug and play script, with minimum required tweaks.
- Fast, yet powerful design features that works great for both tech-savvy and ordinary users.
- Ease of customization. You can customize anything, from Icons to stylesheets.

Blinds is accessible, intuitive, and provides easy way to make a dark mode version for your website's user interface.

## Quick installation and use

### NPM
1. Open your terminal and move to the root directory of your project.
    ```
    cd project
    ```

2. Run the following command to install Blinds in your project:
    ```
    npm install blinds
    ```

3. In your HTML code, create a tag that will hold the toggle icons and give it your prefered style:
    ```
    <div id="blinds_id" style="width: 24px; height: 24px; position: fixed; top: 10px; right: 10px;"><div>
    ```

4. Import Blinds in your JavaScript code:
    ```
    import { blindsInit } from 'blinds';
    ```

5. Initialize Blinds by calling the function:
    ```
    blindsInit();
    ```

6. Activate the toggle function of the icons by adding the event listener:

    ```
    document.querySelector('#blinds_id').onclick = () => {
        blindsToggle();
    }
    ```

### Using script via CDN
1. Create a tag that will hold the toggle icons and give it your prefered style:
    ```
    <div id="blinds_id" style="width: 24px; height: 24px; position: fixed; top: 10px; right: 10px;"></div>
    ```

2. Add Blinds to your HTML through the script tag. Do this preferably just before the closing body tag. The blinds script must be loaded before your script.js (can be named anything really):

    ```
    <script src="https://cdn.jsdelivr.net/gh/Ikennaf1/blinds@latest/dist/blinds.min.js"></script>
    <script src="script.js"></script>
    </body>    
    ```

3. In your script.js file, initialize blinds:
    ```
    blindsInit()
    ```

4. Activate the toggle function of the icons by adding the event listener:

    ```
    document.querySelector('#blinds_id').onclick = () => {
        blindsToggle();
    }
    ```

That's it, you have successfully installed dark mode on your website. Go ahead and click on the icon. Blinds ships with a default 'bare minimum' style. To learn how you can further customize blinds to suit your needs, head over to the [documentation](https://websitecomingsoon.dev).


## License

Blinds is open-sourced software licensed under the [GNU General Public License version 2](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html).


CHANGELOG.md
# Change Log

## [v0.6.6] - 2023-06-26
### Fixed
- Hook: `useDefaultIcons` - Fixed unpredective behavior of the icons by adding getIcons hook

## [v0.6.3] - 2023-06-26
### Fixed
- Hook: `useDefaultIcons` - Fixed unpredective behavior of the icons by adding useDefaultIcons hook

## [v0.6.2] - 2023-05-24
### Fixed
- Hook: `activateDarkMode`

## [v0.6.0] - 2023-05-24
### Changed
- Hooks: `setOnToggle`, `setOnToggleDark`, `setOnToggleNormal` now `onToggle`, `onToggleDark`, `onToggleNormal`

## [v0.5.2] - 2023-05-02
### Added
- New hook: `setOnToggleDark`, `setOnToggleNormal`

## [v0.5.1] - 2023-05-02
### Added
- New hook: `setOnToggle`

## [v0.4.1] - 2023-05-02
### Added
- Users can now set the `id` for the generated style tag, and the `id` for the toggle icons.
- Users can now set the name or key for the generated cookie, or for the local storage.
- Users can now choose between cookie based storage, or local based storage (local storage is still default).

## [v0.4.0] - 2023-05-02
### Added
- Ability to store with local storage, in addition to cookie based storage.
- Local storage is now default storage option.

## [v0.3.2] - 2023-05-01
### Changed
- Abstracted away the storage functions so more storage types can be possible.

## [v0.3.1] - 2023-04-30
### Changed
- Default `id` for icons in the DOM is now `blinds_id`.
- Default `id` for the dark mode `style` tag is now `blinds_style_id`.
- Removed the unnecessary blinds_settings.json file.

## [v0.3.0] - 2023-04-17
### Added
- Hurray! First stable version released!


blinds.js

    var doc = document.createElement('style');
    doc.setAttribute('type', 'text/css');
    doc.setAttribute('id', 'osd_generated');
    change_theme_mode();
    document.head.appendChild(doc);

window.onload = function()
{
    widget_display();
    widget_click();
}

/**
 * This function toggles the color scheme between original and dark mode.
 * Also toggles the widget between normal mode and dark mode.
 * @returns void
 */
function widget_click()
{
    var widget = document.getElementById('osd_blinds_widget_id');
    widget.onclick = function()
    {
        if (osd_blinds_cookie_get()==0 || osd_blinds_cookie_get()==null) {
            osd_blinds_cookie_set();
            change_theme_mode();
            widget_display();
            return;
        }
        if (osd_blinds_cookie_get()=='osd_blinds_cookieMTU5NjM4MTU0OTI0Ng') {
            osd_blinds_cookie_unset();
            change_theme_mode();
            widget_display();
            return;
        }
    }
}

/**
 * Changes the color scheme between normal and dark modes.
 * @returns string
 */
function change_theme_mode() {
    var stylesheet = '\
        @media screen {\
        body, body div,\
        textarea, select,\
        select option,\
        input[type="checkbox"],\
        input[type="radio"],\
        span[role="textbox"],\
        span[role="combobox"],\
        input[type="search"],\
        input[type="number"],\
        input[type="text"],\
        input[type="email"],\
        input[type="tel"],\
        input[type="year"],\
        input[type="month"],\
        input[type="week"],\
        input[type="datetime-local"],\
        input[type="date"],\
        input[type="url"],\
        input[type="password"] {\
            background-color: #000000 !important;\
            color: #a0a0a0 !important;\
        }\
        \
        header, header div,\
        main, main div,\
        table, th, tr, td, table div, table p,\
        article, article div,\
        aside, aside div, form,\
        footer, footer div,\
        h1, h2, h3, h4, h5, h6,\
        section, p, p div, ul li {\
            background-color: #0a0a0a !important;\
            color: #a0a0a0 !important;\
        }\
        \
        aside h1,\
        aside h2,\
        aside h3,\
        aside h4,\
        aside h5,\
        aside h6 {\
            color: #a0a0a0 !important;\
        }\
        \
        #osd_blinds_widget_id {\
            background: none !important;\
        }\
        \
        button, .btn {\
            background-color: #000000 !important;\
            color: currentColor !important;\
        }\
    }\
    @media print {\
        #osd_blinds_widget_id {\
            display: none !important;\
        }\
    }\
    ';

    if (osd_blinds_cookie_get()==0 || osd_blinds_cookie_get()==null) {
        doc.textContent =
        `
        /*style*/
        @media print {
            #osd_blinds_widget_id {
                display: none !important;
            }
        }
        `;
        return;
    }
    if (osd_blinds_cookie_get()=='osd_blinds_cookieMTU5NjM4MTU0OTI0Ng') {
        doc.textContent = stylesheet;
        return;
    }
}

/**
 * The sun SVG.
 * @returns void
 */
function set_to_sun()
{
    var widget = document.getElementById('osd_blinds_widget_id');
    widget.innerHTML = '<a title="Switch to Normal Mode"><svg xmlns="http://www.w3.org/2000/svg" width="36" height="36" viewBox="0 0 24 24" fill="#a0a0a0" stroke="#a0a0a0" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" style="cursor:pointer;" class="feather feather-sun"><circle cx="12" cy="12" r="5"></circle><line x1="12" y1="1" x2="12" y2="3"></line><line x1="12" y1="21" x2="12" y2="23"></line><line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line><line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line><line x1="1" y1="12" x2="3" y2="12"></line><line x1="21" y1="12" x2="23" y2="12"></line><line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line><line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line></svg></a>';
    return;
}

/**
 * The moon SVG.
 * @returns void
 */
function set_to_moon()
{
    var widget = document.getElementById('osd_blinds_widget_id');
    widget.innerHTML = '<a title="Switch to Dark Mode"><svg xmlns="http://www.w3.org/2000/svg" width="36" height="36" viewBox="0 0 24 24" fill="currentColor" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" style="cursor:pointer;" class="feather feather-moon"><path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path></svg></a>';
    return;
}

/**
 * Displays the widget.
 * @returns void
 */
function widget_display()
{
    if (osd_blinds_cookie_get()==0 || osd_blinds_cookie_get()==null) {
        set_to_moon();
        return;
    }

    if (osd_blinds_cookie_get()=='osd_blinds_cookieMTU5NjM4MTU0OTI0Ng') {
        set_to_sun();
        return;
    }
}

/**
 * Sets the cookie!
 * @returns void
 */
function osd_blinds_cookie_set()
{
    document.cookie = "osd_blinds_cookie=osd_blinds_cookieMTU5NjM4MTU0OTI0Ng; path=/;";
    return;
}

/**
 * Checks for the plugin's cookie existence.
 * @param {*} cookie_name
 * @returns string if cookie exists, null if it doesn't.
 */
function osd_blinds_cookie_get(cookie_name='osd_blinds_cookie')
{
    var name = cookie_name + "=";
    var decodedCookie = decodeURIComponent(document.cookie);
    var ca = decodedCookie.split(';');
    for(var i = 0; i <ca.length; i++) {
        var c = ca[i];
        while (c.charAt(0) == ' ') {
            c = c.substring(1);
        }
        if (c.indexOf(name) == 0) {
        return c.substring(name.length, c.length);
        }
    }
  return null;
}

/**
 * Deletes the plugin's cookie, if it is set.
 * @returns void
 */
function osd_blinds_cookie_unset()
{
    document.cookie = "osd_blinds_cookie=; path=/; expires=Thu, 01 Jan 1970 00:00:00 GMT;";
    return;
}


blinds.php
<?php
/**
 * @package Blinds
 * @version 0.0.7
 */

/**
 * Plugin Name: Blinds
 * Plugin URI: https://ikennaf1.github.io/blinds/
 * Description: A plugin that enables WordPress end users view a blog or website in night or dark mode
 * Author: Ike Felix
 * Version: 0.0.7
 * Author URI: ikennaf1.github.io
 * License: GPLv2 or later
 */

/**
 * Runs on activation of the Blinds plugin.
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	exit;
}

function osd_blinds_activation()
{
    	register_activation_hook(__FILE__, 'osd_blinds_activation');
}

/**
 * Runs on plugin deactivation.
 * Deletes any cookies left on browser.
 */
function osd_blinds_deactivation()
{
    	register_deactivation_hook(__FILE__, 'osd_blinds_deactivation');
    	if (isset($_COOKIE['osd_blinds_cookie'])) {
        setcookie('osd_blinds_cookie', '', time()-3600);
        unset($_COOKIE['osd_blinds_cookie']);
    	}
}

function osd_blinds_widget($content)
{
	global $post;
    	$content .= '<div style="position: fixed; top: 3rem; right: 3rem; z-index: 9999999999;" id="osd_blinds_widget_id"></div>';
    	if (is_admin() || is_network_admin()) {
        echo $content;
        return;
    }
	
	if ($post->post_type == 'post-new.php') {
		return $content;
	}
	
	if ($post->post_type == 'product') {
		echo $content;
	}
	
	return $content;
}

/**
 * Hooks the widget (switch or toggle or control) to the WordPress display.
 */
function osd_blinds_filter()
{
    	add_filter('the_content', 'osd_blinds_widget');
    	add_filter('the_excerpt', 'osd_blinds_widget');
	add_filter('woocommerce_before_shop_loop', 'osd_blinds_widget');
	add_filter('woocommerce_before_single_product', 'osd_blinds_widget');

    	if (is_admin() || is_network_admin()) {
    	add_filter('in_admin_header', 'osd_blinds_widget');
    	return;
    }

	return;
}

//wp_enqueue_script('blinds.js', plugins_url('/blinds/blinds.js'), '', filemtime(dirname(__FILE__, 1).'/blinds.js'));
wp_enqueue_script('blinds.js', plugins_url('/blinds/blinds.js'));
osd_blinds_filter();


script.js
import * as blinds from './blinds.min.js';

blinds.setBlindsStyles({
    "*": {
        "color": "hsl(228, 5%, 70%)",
        "border-color": "hsl(228, 20%, 30%)"
    },
    "div#blinds_dark_mode_id": {
        "color": "coral"
    },
    "body": {
        "background-color": "hsl(228, 20%, 10%)",
        "color": "#f7f8fc"
    },
    ".header-text": {
        "color": "coral"
    },
    ".header-text .header-logo": {
        "color": "coral"
    },
    ".header-text h1": {
        "color": "coral"
    },
    ".body-items": {
        "background-color": "hsl(228, 20%, 20%)"
    },
    ".content-items:hover": {
        "background-color": "hsl(228, 20%, 15%)"
    }
});

blinds.setBlindsIcons({
    "dark": "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"currentColor\" class=\"w-6 h-6\"><path d=\"M12 2.25a.75.75 0 01.75.75v2.25a.75.75 0 01-1.5 0V3a.75.75 0 01.75-.75zM7.5 12a4.5 4.5 0 119 0 4.5 4.5 0 01-9 0zM18.894 6.166a.75.75 0 00-1.06-1.06l-1.591 1.59a.75.75 0 101.06 1.061l1.591-1.59zM21.75 12a.75.75 0 01-.75.75h-2.25a.75.75 0 010-1.5H21a.75.75 0 01.75.75zM17.834 18.894a.75.75 0 001.06-1.06l-1.59-1.591a.75.75 0 10-1.061 1.06l1.59 1.591zM12 18a.75.75 0 01.75.75V21a.75.75 0 01-1.5 0v-2.25A.75.75 0 0112 18zM7.758 17.303a.75.75 0 00-1.061-1.06l-1.591 1.59a.75.75 0 001.06 1.061l1.591-1.59zM6 12a.75.75 0 01-.75.75H3a.75.75 0 010-1.5h2.25A.75.75 0 016 12zM6.697 7.757a.75.75 0 001.06-1.06l-1.59-1.591a.75.75 0 00-1.061 1.06l1.59 1.591z\" /></svg>",
    "normal": "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"currentColor\" class=\"w-6 h-6\"><path fill-rule=\"evenodd\" d=\"M9.528 1.718a.75.75 0 01.162.819A8.97 8.97 0 009 6a9 9 0 009 9 8.97 8.97 0 003.463-.69.75.75 0 01.981.98 10.503 10.503 0 01-9.694 6.46c-5.799 0-10.5-4.701-10.5-10.5 0-4.368 2.667-8.112 6.46-9.694a.75.75 0 01.818.162z\" clip-rule=\"evenodd\" /></svg>"
});

blinds.blindsInit();

document.querySelector('#blinds_id').onclick = () => {
    blinds.blindsToggle();
}
